### DISCUSSION TITLE
(small) Improvement for reconnect after "out of range" disconnect

### DISCUSSION THREAD

**Discussion by DrNeuroSurg** [pschatzmann/ESP32-A2DP#142] [Ideas]
I'm very happy with this library ! It does, what I needed.
 
I'm using it for a speaker-like DEVICE, which has no manual controls. The problem is, when I'm connected to my DEVICE and move with my phone out of range, the DEVICE tries to reconnect AUTOCONNECT_TRY_NUM. After that, I have to connect manually my phone to my DEVICE (because of clear_last_connection). That's annoying. 

So I've done this:

Introduced (in BluetoothA2DPSink.h protected:)

bool hasConnection = false;

In BluetoothA2DPSink::av_hdl_a2d_evt I added after

             if (bt_connected!=nullptr){
                    (*bt_connected)();
                }                
                
                hasConnection = true;

and later (same method) I modified

         } else if (a2d->conn_stat.state == ESP_A2D_CONNECTION_STATE_CONNECTING){
                ESP_LOGI(BT_AV_TAG, "ESP_A2D_CONNECTION_STATE_CONNECTING");

                if(!hasConnection) 
                {
                  connection_rety_count++;
                }

The behavior is as expected: If DEVICE ist connected, it tries forever to reconnect. But if you power the DEVICE on and there is no connection (after AUTOCONNECT_TRY_NUM) it is awaiting a  connection for a different phone.

I hope, you know, what I mean and it is helpful for other users.

Keep coding :-)

**Comment by pschatzmann** [comment]
I think you missed the following  https://github.com/pschatzmann/ESP32-A2DP/issues/108
If I get it right, In addition you approach makes only sense when you always connect with the same device.

**Reply by DrNeuroSurg** [reply]
Hmm - #108 is about BluetoothA2DPSource, or am I wrong ?
My DEVICE is BluetoothA2DPSink, to be clear. 
So with my idea - I'm on testing this - my DEVICE (ESP32)  tries at startup (power-on) connect to last_device (if there is any) - if my ESP32 can't connect, after NUM_TRIES, last_device is cleared as intended and so on.
But if my DEVICE (ESP32) has a running connection AND my phone (source) disconnect (for what reason - out of range), my DEVICE (ESP32) tries to connect forever. So if my phone comes into the range it connects.

Or missed I anything to make that easier ? And I can't assign a specific bda, because I don't know it.

**Comment by pschatzmann** [comment]
My mistake I was meaning https://github.com/pschatzmann/ESP32-A2DP/issues/97
I was adding the method set_try_reconnect_max_count which lets you define how long you want to wait until you need to manually reconnect from the phone

**Reply by DrNeuroSurg** [reply]
Yes, I saw this. BUT: If I'm out of range for a long time (grabbing a cup of coffee or so) I have manually initiate a connection from my phone ..
And if I set NUM_TRIE to a very, very large number, I don't have any chance to connect my device with another phone ..
Best solution were to add a button to clear_lastconnection, but I don't have any buttons :-(

BUT: what I could do (in my sketch) , device starts up -> set connection tries to a small number and after a successful connection I have to set it to a very large number ... This should work !

**Comment by pschatzmann** [comment]
I just made some tests and I think it is possible to have both - retries and being connectable from other devices.

**Comment by pschatzmann** [comment]
I committed a small change:  After the 2nd retry I switch on the discoverability again.
Can you double check as well ?

**Reply by DrNeuroSurg** [reply]
YEP - it takes some time

**Reply by DrNeuroSurg** [reply]
It does not work as expected - sorry. After 10 or 12 tries it stops. hmm.

W/o your modification I did this (and it works):

```
#include "BluetoothA2DPSink.h"
#include "limits.h"

BluetoothA2DPSink a2dp_sink;

const char * dev_name = "A2DP_SINK";


void set_retry_count(int num_retries)
{
      Serial.print("SETTING RECONNECT TRIES TO: "); Serial.println(num_retries);
      a2dp_sink.set_try_reconnect_max_count(num_retries);
  
}
void connection_state_changed(esp_a2d_connection_state_t state, void *ptr){

   Serial.println(a2dp_sink.to_str(state));

   // if connected set retry_count to a very, very large number
   
   if(state == ESP_A2D_CONNECTION_STATE_CONNECTED) {

      set_retry_count(INT_MAX);
    }
}

void setup() {
  
  Serial.begin(115200);
  a2dp_sink.set_on_connection_state_changed(connection_state_changed);
  
  Serial.println("POWER ON");

  set_retry_count(3);

  a2dp_sink.start(dev_name);
  Serial.println("STARTED");
  
}

void loop() {
 
}

```

**Comment by pschatzmann** [comment]
The default retry count is 10. You can increase it ..

```
void setup() {
  Serial.begin(115200);  
  set_retry_count(INT_MAX);

  a2dp_sink.start(dev_name);
  Serial.println("STARTED");
  
}

void loop() { 
}
```

**Reply by DrNeuroSurg** [reply]
Tried this also:
 It reconnects if I switch BT off on my phone1 and on again. BUT: I can't connect with another phone. 
It seems to me, the intervals are too short, to get a connection. It's discoverable but not connectable.

**Comment by pschatzmann** [comment]
With which Arduino-ESP version are you testing ?
Wait - I did not test this case. I think that is a different issue, that I noticed. If you switch off bluetooth - I think it should not reconnect. Only if it lost the connection.

Currently if you switch off phone1 and on again it reconnects - and you can not connect from another phone
If you switch off phone1 - and not on again - you should be able to connect from phone2

And I am afraid that this is still another case from the move out of range: switching off is sending a notification - moving out of range not...

**Reply by DrNeuroSurg** [reply]
https://raw.githubusercontent.com/espressif/arduino-esp32/gh-pages/package_esp32_dev_index.json

**Reply by DrNeuroSurg** [reply]
> Currently if you switch off phone1 and on again it reconnects - and you can not connect from another phone
If you switch off phone1 - and not on again - you should be able to connect from phone2

Nope - if I leave phone1 (BT) switched of I can't connect with phone2 - I can "see" it .. that's all

**Comment by pschatzmann** [comment]
I assume you selected 2.0.1 in Tools -> Boards -> Board Manager  - Enter ESP32

**Reply by DrNeuroSurg** [reply]
Yep - 2.0.1 is installed

**Comment by pschatzmann** [comment]
I noticed that aggressive reconnect quite some time ago and it was disturbing me as well. I  took the opportunity to streamline things a bit:
https://github.com/pschatzmann/ESP32-A2DP/wiki/Auto-Reconnect
